server/sonar-server/src/main/java/org/sonar/server/computation/issue/commonrule/CommentDensityRule.java

```diff
@@ -0,0 +1,88 @@
+
+import static java.lang.String.format;
+
+public class CommentDensityRule extends CommonRule {
+
+  private final MeasureRepository measureRepository;
+  private final Metric commentDensityMetric;
+  private final Metric commentLinesMetric;
+  private final Metric nclocMetric;
+
+  public CommentDensityRule(ActiveRulesHolder activeRulesHolder, MeasureRepository measureRepository, MetricRepository metricRepository) {
+    super(activeRulesHolder, CommonRuleKeys.INSUFFICIENT_COMMENT_DENSITY);
+    this.measureRepository = measureRepository;
+    this.commentDensityMetric = metricRepository.getByKey(CoreMetrics.COMMENT_LINES_DENSITY_KEY);
+    this.commentLinesMetric = metricRepository.getByKey(CoreMetrics.COMMENT_LINES_KEY);
+    this.nclocMetric = metricRepository.getByKey(CoreMetrics.NCLOC_KEY);
+  }
+
+  @Override
+  protected CommonRuleIssue doProcessFile(Component file, ActiveRule activeRule) {
+    Optional<Measure> commentDensityMeasure = measureRepository.getRawMeasure(file, commentDensityMetric);
+    Optional<Measure> commentLinesMeasure = measureRepository.getRawMeasure(file, commentLinesMetric);
+    Optional<Measure> nclocMeasure = measureRepository.getRawMeasure(file, nclocMetric);
+
+    if (commentDensityMeasure.isPresent() && nclocMeasure.isPresent() && nclocMeasure.get().getIntValue() > 0) {
+      // this is a small optimization to not load the minimum value when the measures are not present
+      double minCommentDensity = getMinDensity(activeRule);
+      if (commentDensityMeasure.get().getDoubleValue() < minCommentDensity) {
+        return generateIssue(commentDensityMeasure, commentLinesMeasure, nclocMeasure, minCommentDensity);
+      }
+    }
+    return null;
+  }
+
+  private double getMinDensity(ActiveRule activeRule) {
+    double min = getMinDensityParam(activeRule, CommonRuleKeys.INSUFFICIENT_COMMENT_DENSITY_PROPERTY);
+    if (min >= 100.0) {
+      throw new IllegalStateException("Minimum density of rule [" + activeRule.getRuleKey() + "] is incorrect. Got [100] but must be strictly less than 100.");
+    }
+    return min;
+  }
+
+  private CommonRuleIssue generateIssue(Optional<Measure> commentDensityMeasure, Optional<Measure> commentLinesMeasure, Optional<Measure> nclocMeasure, double minCommentDensity) {
+    int commentLines = commentLinesMeasure.isPresent() ? commentLinesMeasure.get().getIntValue() : 0;
+    int ncloc = nclocMeasure.get().getIntValue();
+    int minExpectedCommentLines = (int) Math.ceil(minCommentDensity * ncloc / (100 - minCommentDensity));
+    int missingCommentLines = minExpectedCommentLines - commentLines;
+    if (missingCommentLines <= 0) {
+      throw new IllegalStateException(format("Bug in measures of comment lines - density=%s, comment_lines= %d, ncloc=%d, threshold=%s%%", commentDensityMeasure.get()
+        .getDoubleValue(), commentLines, nclocMeasure.get().getIntValue(), minCommentDensity));
+    }
+
+    // TODO declare min threshold as int but not float ?
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)